File: python/src/ndslive/math/tileid.py
from .morton import MortonCode


class PackedTileId:
    """
    Represents a tile in a hierarchical tiling system.
    Provides methods to extract level, size, and coordinate
    information from the packed tile ID.

    All constructors validate inputs and raise ValueError for invalid tile IDs.
    """
    def __init__(self, value=0):
        self.value = value
        self._validate()
    
    @classmethod
    def from_tile_index(cls, morton_number, level):
        """
        Create a PackedTileId directly from a tile morton number and level.

        This constructs a tile using the morton number as the tile's index
        at the specified level, without any coordinate conversion.

        Args:
            morton_number: The tile's morton number (0 to 2^(2*level+1) - 1)
            level: Tile level (0-15)

        Returns:
            PackedTileId with the specified morton number at the given level

        Raises:
            ValueError: If level or morton_number are out of valid range

        Example:
            >>> tile = PackedTileId.from_tile_index(4, 2)
            >>> tile.morton_number()  # Returns 4
        """
        # Validate level
        if not (0 <= level <= 15):
            raise ValueError(f"Invalid level {level} (must be 0-15)")

        # Validate morton number for this level
        max_morton = (1 << (2 * level + 1)) - 1
        if not (0 <= morton_number <= max_morton):
            raise ValueError(
                f"Invalid morton number {morton_number} for level {level} "
                f"(allowed: 0-{max_morton})"
            )

        value = morton_number + (1 << (16 + level))
        return cls(value)

    @classmethod
    def from_morton_and_level(cls, morton_code, level):
        """
        Create a PackedTileId that contains the point encoded by a MortonCode.

        This method finds the tile at the specified level that contains the
        full-precision NDS coordinates encoded in the MortonCode. The resulting
        tile's morton_number will NOT equal the input morton_code.value() unless
        the point happens to be in that specific tile.

        Note: If you want to create a tile with a specific morton number, use
        from_tile_index() instead.

        Args:
            morton_code: A MortonCode representing full-precision NDS coordinates
            level: Tile level (0-15)

        Returns:
            PackedTileId of the tile containing the encoded point

        Raises:
            ValueError: If level is out of valid range
        """
        # Validate level
        if not (0 <= level <= 15):
            raise ValueError(f"Invalid level {level} (must be 0-15)")

        # Get NDS coordinates from morton code
        x_coord, y_coord = morton_code.to_nds_coordinates()
        
        # Handle negative coordinates
        if x_coord < 0:
            x_coord += (1 << 32)
        
        if y_coord < 0:
            y_coord += (1 << 31)
        
        # Calculate tile coordinates
        n_level = 31 - level
        n_x = x_coord >> n_level
        n_y = y_coord >> n_level
        
        # Create morton code from tile coordinates
        temp = MortonCode.from_nds_coordinates(n_x, n_y)
        
        # Calculate packed tile ID value
        value = temp.value() + (1 << (16 + level))
        
        return cls(value)

    def level(self):
        """
        Level of the tile (0..15)
        """
        level = 0
        tile_id = self.value >> 16
        while tile_id > 1:
            tile_id >>= 1
            level += 1
        return level

    def size(self):
        """
        Size of the tile in NDS coordinate units.
        """
        return 1 << (31 - self.level())

    def center(self):
        """
        Returns the center of the tile in NDS coordinates.
        """
        x, y = self.south_west_corner()
        half_size = self.size() // 2
        return x + half_size, y + half_size

    def south_west_corner(self):
        """
        Returns the south-west corner of the tile in NDS coordinates.
        """
        morton_number = self.morton_number()
        return MortonCode(morton_number << (63 - (2 * self.level() + 1))).to_nds_coordinates()

    def north_east_corner(self):
        """
        Returns the north-east corner of the tile in NDS coordinates.
        """
        x, y = self.south_west_corner()
        size = self.size()
        return x + size, y + size

    def morton_number(self):
        """
        Returns the Morton number of the tile, calculated by subtracting
        the level-specific offset from the packed tile ID value.
        """
        tile_level = self.level()
        return self.value - (1 << (16 + tile_level))

    def _validate(self):
        """
        Validates this PackedTileId.

        Raises:
            ValueError: If the tile ID is invalid, with a detailed error message
        """
        min_packed_tile_id = 1 << 16
        if self.value < min_packed_tile_id:
            raise ValueError(
                f"Invalid PackedTileId({self.value}): value must be >= {min_packed_tile_id}"
            )

        tile_level = self.level()
        morton = self.morton_number()
        max_morton = (1 << (2 * tile_level + 1)) - 1

        if morton < 0 or morton > max_morton:
            raise ValueError(
                f"Invalid PackedTileId({self.value}): morton number {morton} "
                f"exceeds valid range for level {tile_level} (allowed: 0-{max_morton})"
            )

    def _deinterleave_morton(self, morton, level):
        """
        Extract X and Y coordinates from morton number.

        In the NDS tiling system, X has (level+1) bits and Y has level bits,
        creating a rectangular grid that is twice as wide as it is tall.

        Args:
            morton: The morton number to deinterleave
            level: The tile level

        Returns:
            Tuple of (x, y) coordinates
        """
        x = y = 0
        # Y has 'level' bits, X has 'level + 1' bits
        # Morton bits are interleaved as: X0 Y0 X1 Y1 ... X{level} Y{level-1} X{level}
        for i in range(level):
            if morton & (1 << (2 * i)):
                x |= (1 << i)
            if morton & (1 << (2 * i + 1)):
                y |= (1 << i)
        # Extract the extra X bit
        if morton & (1 << (2 * level)):
            x |= (1 << level)
        return x, y

    def _interleave_coords(self, x, y, level):
        """
        Create morton number from X and Y coordinates.

        In the NDS tiling system, X has (level+1) bits and Y has level bits.

        Args:
            x: X coordinate (0 to 2^(level+1) - 1)
            y: Y coordinate (0 to 2^level - 1)
            level: The tile level

        Returns:
            The interleaved morton number
        """
        morton = 0
        # Interleave level bits from each coordinate
        for i in range(level):
            if x & (1 << i):
                morton |= (1 << (2 * i))
            if y & (1 << i):
                morton |= (1 << (2 * i + 1))
        # Add the extra X bit
        if x & (1 << level):
            morton |= (1 << (2 * level))
        return morton

    def west_neighbour(self):
        """
        Returns the tile to the west of this tile at the same level.

        Wraps around at the antimeridian (180° longitude) - going west from
        the westernmost tile returns the easternmost tile at the same level.

        Returns:
            PackedTileId of the western neighbor (with wrapping)
        """
        level = self.level()
        morton = self.morton_number()

        # Deinterleave to get X, Y
        x, y = self._deinterleave_morton(morton, level)

        # Move west (decrement X with wrapping)
        # X has (level + 1) bits, so max value is 2^(level+1) - 1
        max_x = (1 << (level + 1)) - 1
        x = (x - 1) & max_x

        # Reinterleave
        new_morton = self._interleave_coords(x, y, level)

        return PackedTileId.from_tile_index(new_morton, level)

    def east_neighbour(self):
        """
        Returns the tile to the east of this tile at the same level.

        Wraps around at the antimeridian (180° longitude) - going east from
        the easternmost tile returns the westernmost tile at the same level.

        Returns:
            PackedTileId of the eastern neighbor (with wrapping)
        """
        level = self.level()
        morton = self.morton_number()

        x, y = self._deinterleave_morton(morton, level)

        # Move east (increment X with wrapping)
        # X has (level + 1) bits, so max value is 2^(level+1) - 1
        max_x = (1 << (level + 1)) - 1
        x = (x + 1) & max_x

        new_morton = self._interleave_coords(x, y, level)
        return PackedTileId.from_tile_index(new_morton, level)

    def south_neighbour(self):
        """
        Returns the tile to the south of this tile at the same level.

        Wraps around at the south pole - going south from the southernmost
        tile returns the northernmost tile at the same level.

        Returns:
            PackedTileId of the southern neighbor (with wrapping)
        """
        level = self.level()
        morton = self.morton_number()

        x, y = self._deinterleave_morton(morton, level)

        # Move south (decrement Y with wrapping)
        # Y has level bits, so max value is 2^level - 1
        max_y = (1 << level) - 1
        y = (y - 1) & max_y

        new_morton = self._interleave_coords(x, y, level)
        return PackedTileId.from_tile_index(new_morton, level)

    def north_neighbour(self):
        """
        Returns the tile to the north of this tile at the same level.

        Wraps around at the north pole - going north from the northernmost
        tile returns the southernmost tile at the same level.

        Returns:
            PackedTileId of the northern neighbor (with wrapping)
        """
        level = self.level()
        morton = self.morton_number()

        x, y = self._deinterleave_morton(morton, level)

        # Move north (increment Y with wrapping)
        # Y has level bits, so max value is 2^level - 1
        max_y = (1 << level) - 1
        y = (y + 1) & max_y

        new_morton = self._interleave_coords(x, y, level)
        return PackedTileId.from_tile_index(new_morton, level)

    def __str__(self):
        return f"PackedTileId(value={self.value})"

    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return self.value != other.value

    def __lt__(self, other):
        return self.value < other.value

    def __int__(self):
        return self.value

    def __str__(self):
        return f"PackedTileId(value={self.value})"


def get_tile_ids_for_bounding_box(sw_x, sw_y, ne_x, ne_y, level):
    """
    Get all tile IDs that intersect with a bounding box defined by NDS coordinates.
    
    Args:
        sw_x: South-west corner X coordinate (longitude) in NDS coordinates
        sw_y: South-west corner Y coordinate (latitude) in NDS coordinates
        ne_x: North-east corner X coordinate (longitude) in NDS coordinates
        ne_y: North-east corner Y coordinate (latitude) in NDS coordinates
        level: Tile level (0-15)
    
    Returns:
        List of PackedTileId objects that intersect with the bounding box
    """
    tile_ids = []
    
    # Calculate tile size at this level
    tile_size = 1 << (31 - level)
    
    # Calculate tile indices for the bounding box corners
    # We need to handle the coordinate system properly
    start_tile_x = sw_x // tile_size
    start_tile_y = sw_y // tile_size
    end_tile_x = ne_x // tile_size
    end_tile_y = ne_y // tile_size
    
    # Iterate through all tiles in the bounding box
    for tile_y in range(start_tile_y, end_tile_y + 1):
        for tile_x in range(start_tile_x, end_tile_x + 1):
            # Calculate the south-west corner of this tile
            tile_sw_x = tile_x * tile_size
            tile_sw_y = tile_y * tile_size
            
            # Create morton code from the tile's south-west corner
            morton = MortonCode.from_nds_coordinates(tile_sw_x, tile_sw_y)
            
            # Create the packed tile ID
            tile_id = PackedTileId.from_morton_and_level(morton, level)
            tile_ids.append(tile_id)

    return tile_ids


def bounding_box_from_tile_ids(tile_ids):
    """
    Create a tight bounding box from a list of tile IDs.

    This function computes the minimal bounding box in NDS coordinates that
    covers all the specified tiles. The bounding box is guaranteed to be
    tight, meaning it starts at the south-west corner of the westernmost/
    southernmost tile and ends at the north-east corner of the easternmost/
    northernmost tile.

    Important property: When given a single tile ID, the resulting bounding
    box will return only that tile when passed to get_tile_ids_for_bounding_box()
    at the same level.

    Args:
        tile_ids: List of PackedTileId objects or integer tile IDs

    Returns:
        Tuple of (sw_x, sw_y, ne_x, ne_y) in NDS coordinates, representing
        the minimal bounding box that covers all tiles.

    Raises:
        ValueError: If tile_ids list is empty

    Example:
        >>> tile = PackedTileId(545554681)
        >>> bbox = bounding_box_from_tile_ids([tile])
        >>> # bbox will exactly match tile's boundaries
    """
    if not tile_ids:
        raise ValueError("tile_ids list cannot be empty")

    # Ensure we have PackedTileId objects
    tiles = []
    for tid in tile_ids:
        if isinstance(tid, int):
            tiles.append(PackedTileId(tid))
        elif isinstance(tid, PackedTileId):
            tiles.append(tid)
        else:
            raise TypeError(f"Expected int or PackedTileId, got {type(tid)}")

    # Initialize with first tile's bounds
    first_sw_x, first_sw_y = tiles[0].south_west_corner()
    first_ne_x, first_ne_y = tiles[0].north_east_corner()

    min_x = first_sw_x
    min_y = first_sw_y
    max_x = first_ne_x
    max_y = first_ne_y

    # Expand bounds to include all tiles
    for tile in tiles[1:]:
        sw_x, sw_y = tile.south_west_corner()
        ne_x, ne_y = tile.north_east_corner()

        min_x = min(min_x, sw_x)
        min_y = min(min_y, sw_y)
        max_x = max(max_x, ne_x)
        max_y = max(max_y, ne_y)

    # Adjust NE corner: north_east_corner() returns exclusive boundary (first point outside tile)
    # but get_tile_ids_for_bounding_box() treats coordinates as inclusive.
    # Subtract 1 to get the last point inside the tile rather than first point outside.
    return (min_x, min_y, max_x - 1, max_y - 1)

File: go/tileid.go
// SPDX-License-Identifier: BSD-3-Clause

package ndslivemath

import (
	"fmt"
	"math"
)

// PackedTileId represents a tile in the hierarchical NDS.Live tiling system.
//
// Per the NDS.Live standard, tile IDs are signed 32-bit integers. For levels
// 0-14 the values are positive; for level 15 the values are negative
// (-2147483648 to -1) because the level bit (bit 31) coincides with the sign
// bit of a signed int32.
//
// Internally the value is stored as an unsigned 32-bit integer so that all bit
// operations are free of sign-bit complications. Conversion between signed and
// unsigned happens only at the API boundary: constructors accept either
// representation, and Value returns the signed int32 mandated by the standard.
type PackedTileId struct {
	value uint32 // internal unsigned storage
}

// NewPackedTileId constructs a PackedTileId from a signed int32 tile ID value.
//
// Per the NDS.Live standard, level 15 tiles have negative values
// (-2147483648 to -1) while levels 0-14 have positive values. This accepts the
// signed representation directly; for unsigned input use
// NewPackedTileIdFromUnsigned.
//
// It returns an error if the resulting tile ID is invalid.
func NewPackedTileId(value int32) (PackedTileId, error) {
	// Convert signed int32 to unsigned storage. uint32(value) performs the
	// two's-complement reinterpretation, so e.g. -2147483648 -> 2147483648.
	t := PackedTileId{value: uint32(value)}
	if err := t.validate(); err != nil {
		return PackedTileId{}, err
	}
	return t, nil
}

// NewPackedTileIdFromUnsigned constructs a PackedTileId from an unsigned tile
// ID value. Values outside the 32-bit range are masked to 32 bits, mirroring
// the Python constructor. It returns an error if the tile ID is invalid.
func NewPackedTileIdFromUnsigned(value uint64) (PackedTileId, error) {
	t := PackedTileId{value: uint32(value & 0xFFFFFFFF)}
	if err := t.validate(); err != nil {
		return PackedTileId{}, err
	}
	return t, nil
}

// PackedTileIdFromValue constructs a PackedTileId from the signed NDS.Live
// public value.
func PackedTileIdFromValue(value int32) (PackedTileId, error) {
	return NewPackedTileId(value)
}

// PackedTileIdFromTileIndex creates a PackedTileId directly from a tile morton
// number and level, without any coordinate conversion.
//
// mortonNumber must be in [0, 2^(2*level+1) - 1] and level in [0, 15];
// otherwise an error is returned.
func PackedTileIdFromTileIndex(mortonNumber uint32, level int) (PackedTileId, error) {
	if level < 0 || level > 15 {
		return PackedTileId{}, fmt.Errorf("invalid level %d (must be 0-15)", level)
	}
	maxMorton := (uint64(1) << (2*level + 1)) - 1
	if uint64(mortonNumber) > maxMorton {
		return PackedTileId{}, fmt.Errorf(
			"invalid morton number %d for level %d (allowed: 0-%d)",
			mortonNumber, level, maxMorton)
	}
	value := uint64(mortonNumber) + (uint64(1) << (16 + level))
	t := PackedTileId{value: uint32(value)}
	if err := t.validate(); err != nil {
		return PackedTileId{}, err
	}
	return t, nil
}

// PackedTileIdFromTileXY creates a PackedTileId from tile-grid coordinates at
// the given level. X is in [0, 2^(level+1)-1], Y is in [0, 2^level-1].
// Coordinates use the NDS Morton tile-grid order and are inverse to X and Y.
func PackedTileIdFromTileXY(x, y uint32, level int) (PackedTileId, error) {
	if level < 0 || level > 15 {
		return PackedTileId{}, fmt.Errorf("invalid level %d (must be 0-15)", level)
	}
	maxX := (uint32(1) << (level + 1)) - 1
	maxY := (uint32(1) << level) - 1
	if x > maxX || y > maxY {
		return PackedTileId{}, fmt.Errorf(
			"invalid tile coordinates (%d, %d) for level %d (allowed x: 0-%d, y: 0-%d)",
			x, y, level, maxX, maxY)
	}
	return PackedTileIdFromTileIndex(interleaveCoords(x, y, level), level)
}

// PackedTileIdFromNdsCoordinates returns the tile at level that contains the
// given NDS integer coordinate.
func PackedTileIdFromNdsCoordinates(x, y int32, level int) (PackedTileId, error) {
	return PackedTileIdFromMortonAndLevel(MortonFromNdsCoordinates(x, y), level)
}

// PackedTileIdFromWgs84 returns the tile at level that contains the given
// WGS84 coordinate.
func PackedTileIdFromWgs84(longitude, latitude float64, level int) (PackedTileId, error) {
	x, y := NewWgs84(longitude, latitude, 0).ToNdsCoordinates()
	return PackedTileIdFromNdsCoordinates(x, y, level)
}

// PackedTileIdFromMortonAndLevel creates the PackedTileId of the tile at the
// given level that contains the full-precision NDS point encoded by mortonCode.
//
// The resulting tile's MortonNumber will NOT generally equal
// mortonCode.Value(); use PackedTileIdFromTileIndex if you need a specific
// morton number. level must be in [0, 15].
func PackedTileIdFromMortonAndLevel(mortonCode MortonCode, level int) (PackedTileId, error) {
	if level < 0 || level > 15 {
		return PackedTileId{}, fmt.Errorf("invalid level %d (must be 0-15)", level)
	}

	xCoord, yCoord := mortonCode.ToNdsCoordinates()

	// Promote to unsigned magnitudes by adding the wrap-around offsets for
	// negative coordinates (matching the Python reference).
	xu := int64(xCoord)
	yu := int64(yCoord)
	if xu < 0 {
		xu += 1 << 32
	}
	if yu < 0 {
		yu += 1 << 31
	}

	nLevel := 31 - level
	nX := xu >> nLevel
	nY := yu >> nLevel

	temp := MortonFromNdsCoordinates(int32(nX), int32(nY))

	value := temp.Value() + (uint64(1) << (16 + level))
	t := PackedTileId{value: uint32(value)}
	if err := t.validate(); err != nil {
		return PackedTileId{}, err
	}
	return t, nil
}

// Value returns the tile ID as a signed int32 per the NDS.Live standard.
// Level 15 tiles return negative values; levels 0-14 return positive values.
func (t PackedTileId) Value() int32 {
	// int32(uint32) reinterprets the bits, so values with bit 31 set become
	// negative exactly as the standard requires.
	return int32(t.value)
}

// Level returns the level of the tile (0..15).
func (t PackedTileId) Level() int {
	level := 0
	tileID := t.value >> 16
	for tileID > 1 {
		tileID >>= 1
		level++
	}
	return level
}

// Size returns the tile's edge length in NDS coordinate units.
func (t PackedTileId) Size() uint32 {
	return uint32(1) << (31 - t.Level())
}

// MortonNumber returns the tile's Morton number (the value with the
// level-specific offset removed).
func (t PackedTileId) MortonNumber() uint32 {
	tileLevel := t.Level()
	return t.value - (uint32(1) << (16 + tileLevel))
}

// X returns the tile-grid X coordinate at this tile's level.
func (t PackedTileId) X() uint32 {
	x, _ := deinterleaveMorton(t.MortonNumber(), t.Level())
	return x
}

// Y returns the tile-grid Y coordinate at this tile's level.
func (t PackedTileId) Y() uint32 {
	_, y := deinterleaveMorton(t.MortonNumber(), t.Level())
	return y
}

// Center returns the center of the tile in NDS coordinates.
//
// Returned as int64 because, although the center always fits in int32 for
// valid tiles, the corner arithmetic it derives from can exceed int32.
func (t PackedTileId) Center() (int64, int64) {
	x, y := t.SouthWestCorner()
	halfSize := int64(t.Size() / 2)
	return x + halfSize, y + halfSize
}

// PackedTileIdWgs84FromNdsCoordinates converts NDS integer coordinates to
// lon/lat degrees without WGS84 edge normalization.
func PackedTileIdWgs84FromNdsCoordinates(x, y int64) (float64, float64) {
	return float64(x) * 360.0 / math.Exp2(32), float64(y) * 180.0 / math.Exp2(31)
}

// CenterWgs84 returns the center of the tile in lon/lat degrees.
func (t PackedTileId) CenterWgs84() (float64, float64) {
	x, y := t.Center()
	return PackedTileIdWgs84FromNdsCoordinates(x, y)
}

// SouthWestWgs84 returns the south-west tile corner in lon/lat degrees.
func (t PackedTileId) SouthWestWgs84() (float64, float64) {
	x, y := t.SouthWestCorner()
	return PackedTileIdWgs84FromNdsCoordinates(x, y)
}

// NorthEastWgs84 returns the exclusive north-east tile corner in lon/lat
// degrees.
func (t PackedTileId) NorthEastWgs84() (float64, float64) {
	x, y := t.NorthEastCorner()
	return PackedTileIdWgs84FromNdsCoordinates(x, y)
}

// Wgs84Size returns the tile width/height in lon/lat degrees.
func (t PackedTileId) Wgs84Size() (float64, float64) {
	tileSize := float64(t.Size())
	return tileSize * 360.0 / math.Exp2(32), tileSize * 180.0 / math.Exp2(31)
}

// SouthWestCorner returns the south-west (inclusive) corner of the tile in NDS
// coordinates. Returned as int64 for a uniform corner API; SW values always
// fit in int32.
func (t PackedTileId) SouthWestCorner() (int64, int64) {
	mortonNumber := uint64(t.MortonNumber())
	shift := uint(63 - (2*t.Level() + 1))
	x, y := NewMortonCode(mortonNumber << shift).ToNdsCoordinates()
	return int64(x), int64(y)
}

// NorthEastCorner returns the north-east (EXCLUSIVE) corner of the tile in NDS
// coordinates: the first coordinate outside the tile, i.e. SW + size.
//
// Returned as int64 because for low levels (e.g. level 0) the exclusive corner
// is 2^31, which exceeds the signed int32 range.
func (t PackedTileId) NorthEastCorner() (int64, int64) {
	x, y := t.SouthWestCorner()
	size := int64(t.Size())
	return x + size, y + size
}

// DimensionsInMeters returns the tile's (width, height) in meters, computed at
// the tile's center latitude. Width varies with cos(latitude); height is
// constant.
func (t PackedTileId) DimensionsInMeters() (float64, float64) {
	centerX, centerY := t.Center()
	centerWgs := Wgs84FromNdsCoordinates(int32(centerX), int32(centerY))
	tileSize := float64(t.Size())
	return NdsDistanceToMeters(tileSize, tileSize, centerWgs.Lat)
}

// validate checks this PackedTileId's internal (unsigned) value. Error
// messages report the signed API value for clarity.
func (t PackedTileId) validate() error {
	const minPackedTileID = uint32(1) << 16
	if t.value < minPackedTileID {
		return fmt.Errorf(
			"invalid PackedTileId(%d): value must be >= %d or negative for level 15",
			t.Value(), minPackedTileID)
	}

	tileLevel := t.Level()
	morton := uint64(t.MortonNumber())
	maxMorton := (uint64(1) << (2*tileLevel + 1)) - 1
	if morton > maxMorton {
		return fmt.Errorf(
			"invalid PackedTileId(%d): morton number %d exceeds valid range for level %d (allowed: 0-%d)",
			t.Value(), morton, tileLevel, maxMorton)
	}
	return nil
}

// deinterleaveMorton extracts X and Y coordinates from a tile morton number.
// X has (level+1) bits and Y has level bits.
func deinterleaveMorton(morton uint32, level int) (uint32, uint32) {
	var x, y uint32
	for i := 0; i < level; i++ {
		if morton&(1<<(2*i)) != 0 {
			x |= 1 << i
		}
		if morton&(1<<(2*i+1)) != 0 {
			y |= 1 << i
		}
	}
	if morton&(1<<(2*level)) != 0 {
		x |= 1 << level
	}
	return x, y
}

// interleaveCoords builds a tile morton number from X and Y coordinates.
// X has (level+1) bits and Y has level bits.
func interleaveCoords(x, y uint32, level int) uint32 {
	var morton uint32
	for i := 0; i < level; i++ {
		if x&(1<<i) != 0 {
			morton |= 1 << (2 * i)
		}
		if y&(1<<i) != 0 {
			morton |= 1 << (2*i + 1)
		}
	}
	if x&(1<<level) != 0 {
		morton |= 1 << (2 * level)
	}
	return morton
}

// Neighbour returns the same-level tile at a relative grid offset, with
// wrap-around at the respective limits.
func (t PackedTileId) Neighbour(dx, dy int) PackedTileId {
	return t.neighbour(dx, dy)
}

// Neighbor is an American-English alias for Neighbour.
func (t PackedTileId) Neighbor(dx, dy int) PackedTileId {
	return t.Neighbour(dx, dy)
}

// neighbour returns the same-level neighbour after applying dx to X and dy to
// Y, with wrap-around at the respective limits.
func (t PackedTileId) neighbour(dx, dy int) PackedTileId {
	level := t.Level()
	morton := t.MortonNumber()
	x, y := deinterleaveMorton(morton, level)

	if dx != 0 {
		maxX := (uint32(1) << (level + 1)) - 1
		x = uint32(int(x)+dx) & maxX
	}
	if dy != 0 {
		maxY := (uint32(1) << level) - 1
		y = uint32(int(y)+dy) & maxY
	}

	newMorton := interleaveCoords(x, y, level)
	// PackedTileIdFromTileIndex cannot fail here: level and newMorton are
	// within range by construction, so the error is ignored deliberately.
	tile, _ := PackedTileIdFromTileIndex(newMorton, level)
	return tile
}

// WestNeighbour returns the tile to the west at the same level, wrapping at the
// antimeridian.
func (t PackedTileId) WestNeighbour() PackedTileId { return t.neighbour(-1, 0) }

// EastNeighbour returns the tile to the east at the same level, wrapping at the
// antimeridian.
func (t PackedTileId) EastNeighbour() PackedTileId { return t.neighbour(+1, 0) }

// SouthNeighbour returns the tile to the south at the same level, wrapping at
// the south pole.
func (t PackedTileId) SouthNeighbour() PackedTileId { return t.neighbour(0, -1) }

// NorthNeighbour returns the tile to the north at the same level, wrapping at
// the north pole.
func (t PackedTileId) NorthNeighbour() PackedTileId { return t.neighbour(0, +1) }

// String implements fmt.Stringer.
func (t PackedTileId) String() string {
	return fmt.Sprintf("PackedTileId(value=%d)", t.Value())
}

// floorDiv performs floor division (rounding toward negative infinity), which
// differs from Go's built-in integer division (which truncates toward zero)
// for negative operands. divisor is assumed positive (tile sizes always are).
//
// This is critical for parity with Python's // operator used in
// GetTileIdsForBoundingBox.
func floorDiv(a, b int64) int64 {
	q := a / b
	if (a%b != 0) && ((a < 0) != (b < 0)) {
		q--
	}
	return q
}

// GetTileIdsForBoundingBox returns all tile IDs that intersect the bounding box
// defined by the south-west and north-east corners in NDS coordinates, at the
// given level.
//
// Floor division is used for the corner-to-tile-index mapping (matching
// Python's // operator), which matters for negative coordinates.
func GetTileIdsForBoundingBox(swX, swY, neX, neY int32, level int) []PackedTileId {
	var tileIDs []PackedTileId

	tileSize := int64(1) << (31 - level)

	startTileX := floorDiv(int64(swX), tileSize)
	startTileY := floorDiv(int64(swY), tileSize)
	endTileX := floorDiv(int64(neX), tileSize)
	endTileY := floorDiv(int64(neY), tileSize)

	for tileY := startTileY; tileY <= endTileY; tileY++ {
		for tileX := startTileX; tileX <= endTileX; tileX++ {
			tileSwX := tileX * tileSize
			tileSwY := tileY * tileSize

			morton := MortonFromNdsCoordinates(int32(tileSwX), int32(tileSwY))
			// from-morton-and-level cannot fail for a valid level; ignore err.
			tileID, _ := PackedTileIdFromMortonAndLevel(morton, level)
			tileIDs = append(tileIDs, tileID)
		}
	}

	return tileIDs
}

// BoundingBoxFromTileIds computes the minimal NDS bounding box covering all the
// given tiles. It returns (minX, minY, maxX, maxY) where the NE corner has been
// decremented by 1 so that the box is inclusive (the NorthEastCorner of a tile
// is exclusive). It returns an error if tiles is empty.
//
// Values are int64 because the intermediate (exclusive) maxima may exceed
// int32; the decremented results returned here are always within int32 range
// for valid tiles. A single-tile box round-trips: passing the result back
// through GetTileIdsForBoundingBox at the same level yields exactly that tile.
func BoundingBoxFromTileIds(tiles []PackedTileId) (int64, int64, int64, int64, error) {
	if len(tiles) == 0 {
		return 0, 0, 0, 0, fmt.Errorf("tiles list cannot be empty")
	}

	minX, minY := tiles[0].SouthWestCorner()
	maxX, maxY := tiles[0].NorthEastCorner()

	for _, tile := range tiles[1:] {
		swX, swY := tile.SouthWestCorner()
		neX, neY := tile.NorthEastCorner()
		if swX < minX {
			minX = swX
		}
		if swY < minY {
			minY = swY
		}
		if neX > maxX {
			maxX = neX
		}
		if neY > maxY {
			maxY = neY
		}
	}

	return minX, minY, maxX - 1, maxY - 1, nil
}
